### ISSUE TITLE
Dead connections not removed from pool

### ISSUE THREAD

**Issue by ctlajoie** [sidorares/node-mysql2#325]
I'm experiencing a problem where connections that the server has dropped (due to mysql server restart or hitting the wait_timeout) do not get removed from the pool. I wrote a short script that reproduces the problem.

The script opens 5 pooled connections and then waits for something to happen. I set `wait_timeout = 30` for testing. When the timeout is hit, the errors are displayed and then it shows you how many connections are in the pool. It always stays at 5, even after the connections are dropped. I also get the same result if I just restart mysqld after it opens the connections.

I am using mysql2 1.0.0-rc.4

```javascript
var mysql = require('mysql2');
var pool = mysql.createPool('mysql://localhost/test?connectionLimit=5');
//connectionLimit param in the uri doesn't work.. have to set it manually
pool.config.connectionLimit = 5;

pool.reportStatus = function() {
	console.log('Pool connections: %d', this._allConnections.length);
};

pool.on('connection', function(conn) {
	console.log('db connection opened');
	conn.on('error', function(err) {
		console.error(err);
		setTimeout(() => pool.reportStatus(), 3000);
	});
});

// open connections
for (var i=0; i<10; i++)
	pool.query('select 1', ()=>{});

// show connections are opened
setTimeout(() => pool.reportStatus(), 500);


// don't let the script die
var server = require('http').createServer();
server.listen(5151, ()=>{});
```

**Comment by ctlajoie** [comment]
BTW this is the output I get from running the script.

```
db connection opened
db connection opened
db connection opened
db connection opened
db connection opened
Pool connections: 5
{ Error: Connection lost: The server closed the connection.
    at Socket.<anonymous> (/home/chris/dev/ett-mail/node_modules/mysql2/lib/connection.js:110:15)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }
{ Error: Connection lost: The server closed the connection.
    at Socket.<anonymous> (/home/chris/dev/ett-mail/node_modules/mysql2/lib/connection.js:110:15)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }
{ Error: Connection lost: The server closed the connection.
    at Socket.<anonymous> (/home/chris/dev/ett-mail/node_modules/mysql2/lib/connection.js:110:15)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }
{ Error: Connection lost: The server closed the connection.
    at Socket.<anonymous> (/home/chris/dev/ett-mail/node_modules/mysql2/lib/connection.js:110:15)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }
{ Error: Connection lost: The server closed the connection.
    at Socket.<anonymous> (/home/chris/dev/ett-mail/node_modules/mysql2/lib/connection.js:110:15)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }
Pool connections: 5
Pool connections: 5
Pool connections: 5
Pool connections: 5
Pool connections: 5
```

**Comment by sidorares** [comment]
thanks @ctlajoie . I'll try to prepare test with fake server and see if I can predictably reproduce this

**Comment by sidorares** [comment]
in master and published as v1.0.0-rc.5, thanks!
